Wrappers/Python/ccpi/optimisation/functions/FGP_TV.py

```diff
@@ -0,0 +1,275 @@
+    .. math::  \min_{x} \alpha TV(x) + \frac{1}{2}||x-b||^{2}_{2}
+                
+    Parameters:
+      
+      :param domain: Domain of the reconstruction
+      :param regularising_parameter: TV regularising parameter 
+      :param inner_iterations: Iterations of FGP algorithm
+      :param tolerance: Stopping criterion (Default=1e-4)
+      :param nonnegativity: 
+      
+    Reference:
+      
+        Beck, A. and Teboulle, M., 2009. Fast Gradient-Based Algorithms for Constrained
+        Total Variation Image Denoising and Deblurring Problems 
+        IEEE TRANSACTIONS ON IMAGE PROCESSING, VOL. 18, NO. 11, NOVEMBER 2009.
+        
+    '''    
+    
+    
+    def __init__(self, domain, regularising_parameter, inner_iterations, tolerance, nonnegativity):
+        
+        self.regularising_parameter = regularising_parameter
+        self.inner_iterations = inner_iterations
+        self.tolerance = tolerance
+        self.nonnegativity = nonnegativity                
+        self.gradient = Gradient(domain)
+        self.TV = regularising_parameter * MixedL21Norm()                 
+            
+    def __call__(self,x):
+        
+        # evaluate objective function of TV gradient
+        return self.TV(self.gradient.direct(x))
+    
+    
+    def projection_C(self, x, out=None):   
+
+        # This can be replaced by IndicatorBox(lower=0), but is less effient.
+                        
+        if out is None:            
+            if self.nonnegativity:
+                return x.maximum(0.)
+            else:
+                return x
+        else:
+            if self.nonnegativity:            
+                x.maximum(0., out=out)                
+            else:
+                out.fill(x)
+
+    def projection_P(self, x, out=None):
+
+        # This can be replaced by the proximal conjugate of MixedL21Norm, but is less effient.
+        
+        res1 = functools.reduce(lambda a,b: a + b*b, x.containers, x.get_item(0) * 0 )
+        res1.sqrt(out=res1)	
+        res1.maximum(1.0, out=res1)	    
+
+        if out is None:
+            return x.divide(res1)
+        else:
+            x.divide(res1, out=out)
+    
+    
+    def proximal(self, b, tau, out = None):
+                
+        # initialise
+        t = 1        
+        tmp_p = self.gradient.range_geometry().allocate()  
+        tmp_q = tmp_p.copy()
+        x = self.gradient.domain_geometry().allocate()     
+        p1 = tmp_p.copy()
+
+        for k in range(self.inner_iterations):
+                                                                                   
+            t0 = t
+            self.gradient.adjoint(tmp_q, out = x)
+            
+            # this can be replaced by axpby
+            x *= -self.regularising_parameter
+            x *= tau
+            x += b
+            self.projection_C(x, out = x)                       
+
+            self.gradient.direct(x, out=p1)
+            p1 *= 1./(8*self.regularising_parameter)
```

Comment:
Is the current implementation only for 2D TV? I wonder if the value "8" here is specific to 2D?  If specific to 2D, maybe we should name the function FGP_TV2D? If more general than 2D, maybe we should mention in the comments in the beginning that it works for other Ds as well, 1D, 3D, 4D, ...?